### PR TITLE
feat: add support for baseUrl

### PR DESCRIPTION
**Pull request by aslafy-z**
### Background 

I want to host this microservice demo frontend on a subpath (`X.X.X.X/online-boutique/`). I made my reverse proxy rewrite the url by removing `/online-boutique` from the paths it sees. However, the frontend html pages uses absolute urls which cannot be solved with the reverse proxy. 
### Change Summary

- Introduce baseUrl from BASE_URL environment variable
- Add baseUrl to default template variables 
- Prefix all absolute urls with baseUrl
- Update handlers to use baseUrl
- Update redirects to use baseUrl

  
### Additional Notes

Image is available at `docker.io/zadki3l/microservice-demo-frontend:v0.11.0-dev.4`

### Testing Procedure

- Deploy app with `BASE_URL=/online-boutique` environment variable.
  <details>
  <summary>Kustomize example</summary>
  
  ```yaml
	kind: Kustomization
	namespace: online-boutique
	resources:
	- https://github.com/GoogleCloudPlatform/microservices-demo//kustomize/base?ref=6af73dcf7fe1ca653d95ae7a29e12faf8cf02964
	- namespace.yaml
	- ingress.yaml
	components:
	- https://github.com/GoogleCloudPlatform/microservices-demo//kustomize/components/non-public-frontend?ref=6af73dcf7fe1ca653d95ae7a29e12faf8cf02964
	- https://github.com/GoogleCloudPlatform/microservices-demo//kustomize/components/without-loadgenerator?ref=6af73dcf7fe1ca653d95ae7a29e12faf8cf02964
	images:
	 - name: gcr.io/google-samples/microservices-demo/frontend
	   newName: docker.io/zadki3l/microservice-demo-frontend
	   newTag: v0.11.0
	patches:
	- patch: |-
	    - op: add
	      path: "/spec/template/spec/containers/0/env/-"
	      value:
	        name: BASE_URL
	        value: "/online-boutique"
	  target:
	    kind: Deployment
	    name: frontend
  ```
  
  </details>
- Deploy an ingress with `/online-boutique` prefix.
- Validate that everything shows correctly

### CODE REVIEW

File: src/frontend/main.go [hunk 1]
```diff
@@ -141,20 +146,20 @@ func main() {
...
+	r.HandleFunc(baseUrl + "/", svc.homeHandler).Methods(http.MethodGet, http.MethodHead)
+	r.HandleFunc(baseUrl + "/product/{id}", svc.productHandler).Methods(http.MethodGet, http.MethodHead)
+	r.HandleFunc(baseUrl + "/cart", svc.viewCartHandler).Methods(http.MethodGet, http.MethodHead)
+	r.HandleFunc(baseUrl + "/cart", svc.addToCartHandler).Methods(http.MethodPost)
+	r.HandleFunc(baseUrl + "/cart/empty", svc.emptyCartHandler).Methods(http.MethodPost)
+	r.HandleFunc(baseUrl + "/setCurrency", svc.setCurrencyHandler).Methods(http.MethodPost)
+	r.HandleFunc(baseUrl + "/logout", svc.logoutHandler).Methods(http.MethodGet)
+	r.HandleFunc(baseUrl + "/cart/checkout", svc.placeOrderHandler).Methods(http.MethodPost)
+	r.HandleFunc(baseUrl + "/assistant", svc.assistantHandler).Methods(http.MethodGet)
+	r.PathPrefix(baseUrl + "/static/").Handler(http.StripPrefix("/static/", http.FileServer(http.Dir("./static/"))))
```
Comment (aslafy-z) [inline src/frontend/main.go:158]:
```suggestion
	r.PathPrefix(baseUrl + "/static/").Handler(http.StripPrefix(baseUrl + "/static/", http.FileServer(http.Dir("./static/"))))
```
```suggestion
	r.PathPrefix(baseUrl + "/static/").Handler(http.StripPrefix("/static/", http.FileServer(http.Dir("./static/"))))
```

**Review by NimJay — CHANGES_REQUESTED** [review]
Thanks for this feature.

It looks like the "You May Also Like" section of each product page is broken in the staging URL:
Screenshot 2024-06-27 at 2 48 12 PM

Do you know if it might be due to a change in the .html templates?

**Comment by aslafy-z** [comment]
@NimJay I don't think so, the same changes were done on the product listing page without issue: https://github.com/GoogleCloudPlatform/microservices-demo/pull/2609/files#diff-486e203835857587dcd0bd356e4778094d7514a853fcfb8d83b6c0163b471eaeL48-L49.

**Comment by aslafy-z** [comment]
I updated the code to fix Ad and Recommendation templates. I had to pass them the global context. I'm not a fan of that solution, however, it works well.
I pushed an updated image at `docker.io/zadki3l/microservice-demo-frontend:v0.11.0-dev.4`.

**Review by NimJay — APPROVED** [review]

**Pull request merged**